### PR TITLE
Improve word doc formatting

### PR DESCRIPTION
**Pull request by kelliemac**
## Description

Update word-styles-reference.docx for more professional-looking word document formatting, and switch to use of flextable package for table generation to ensure tables look good (and consistent) in both PDF and Word reports.

## Related Issues

Related to https://github.com/FredHutch/VISCtemplates/issues/120

## Checklist

- [ ] This PR includes unit tests
- [ ] This PR establishes a new function or updates parameters in an existing function
  - [ ]  The roxygen skeleton for this function has been updated using `devtools::document`
- [ ] I have updated NEWS.md to describe the proposed changes

### CODE REVIEW

**Comment by kelliemac** [comment]
making good progress on this! outstanding issues as of now:

1. Reproducibility software session information table runs off the page (too wide) in PDF output
2. Study schema table caption says something like "file15c072f288e5" instead of "VDC nnn"
3. No capability to handle short versions of table captions. Consider putting extra details in table footnotes instead.
4. No list of figures or list of tables in Word output. Is this even possible?

**Comment by slager** [comment]
"file15c072f288e5 study schema" in the table caption is expected I think, because that's the actual randomly generated name of the visc project folder created during the test knitting.

**Comment by slager** [comment]
I haven't looked at this carefully yet, but took a glance and one thing I think is to replace the word template header logos with the current VISC and SCHARP source images in the repo just to make sure they're the newest.  The VISC logo, at least, looks like the larger, older one.

**Comment by mayerbry** [comment]
imo, we need larger buy in from the analysis team before we switch between `'kableExtra` and `flextable`. Currently, `kable` is a backbone of all of our reports and potentially part of our training materials. Switching the template might be abrupt without a bit of context and motivation.

This could include a short training (maybe a TIoT), examples/vignette of transitioning table code (I've noticed that there are different functions that do the same thing) so we avoid projects that have cross-dependencies/mixed table formats due to copy and pasting of working tables, and confirming that flextable can re-generate some of our recent tables that may not follow typical formatting (I can check a few of my PK ones). 

VISCfunctions has functions that support input into kables, and so we should also identify what needs to be updated there.

**Comment by kelliemac** [comment]
> imo, we need larger buy in from the analysis team before we switch between `'kableExtra` and `flextable`. Currently, `kable` is a backbone of all of our reports and potentially part of our training materials. Switching the template might be abrupt without a bit of context and motivation.
> 
> This could include a short training (maybe a TIoT), examples/vignette of transitioning table code (I've noticed that there are different functions that do the same thing) so we avoid projects that have cross-dependencies/mixed table formats due to copy and pasting of working tables, and confirming that flextable can re-generate some of our recent tables that may not follow typical formatting (I can check a few of my PK ones).
> 
> VISCfunctions has functions that support input into kables, and so we should also identify what needs to be updated there.

@mayerbry yes I agree that this is a non-trivial change and we need to have some conversations and training before we merge it in! I already set up a TioT for 8/8 to talk about the advantages of moving to flextable and get feedback from the group on potential challenges that might come up. Can you post here an example or two of complicated kable formatting you've used in a previous project, so that I can make sure flextable can cover all of those needs?

I know `VISCfunctions::pretty_pvalues()` is one function that supports input into kables. Can you think of any other examples?

**Comment by kelliemac** [comment]
> I haven't looked at this carefully yet, but took a glance and one thing I think is to replace the word template header logos with the current VISC and SCHARP source images in the repo just to make sure they're the newest. The VISC logo, at least, looks like the larger, older one.

thanks @slager. I forgot that you had changed the VISC logo too, so that was the only one I didn't change. just fixed it.

**Comment by slager** [comment]
I agree with this need to think hard about it. This is amazing for these tables, yet I still find myself thinking about my work on QC reports, where I've seen there's often a need to customize a table pretty extensively to meet various stakeholder needs. I've already seen a lot of edge cases with tables just in the several QC reports I've done, like whether it's extra wide, how many levels of columns we're collapsing rows on, whether the table occupies more than 1 page, whether we want certain groups of rows to stay together when they get split across pages, if there are non-ASCII characters in labels, adding another caption, etc.  That has often required custom workarounds to get the right data visualization.

It's hard to test for all the edge cases for a function which has many potentially interacting arguments and whose output is a side effect that needs to be evaluated visually.  With kableExtra, there's documentation to hopefully find a solution, but if we make our own big function, it will be on us, and potentially harder to customize when we need to.  So definitely some pros and cons.

Looking forward to the TIOT!

**Comment by kelliemac** [comment]
Just to clarify in response to one of your comments @slager: we are no longer creating a custom VISCtemplates table-generating function. We are just encouraging people to use the flextable package instead of kable and kableExtra, so they should still be able to look up the flextable documentation for additional capabilities or issues, rather than turning to us as the first line of defense.

**Comment by kelliemac** [comment]
In my experience, flextable seems to have all of the key capabilities that are present in kableExtra. I'm sure there are some cases where they don't perfectly line up, but for the most part they just use different syntax to achieve the same kinds of effects. I can start drafting a list of kableExtra formatting calls and their equivalent flextable calls.

here is the flextable website for reference: https://davidgohel.github.io/flextable/

**Comment by slager** [comment]
Oops, I hadn't noticed yet that it is no longer a custom function, so ignore the comments of mine related to that!

**Comment by mayerbry** [comment]
Looking forward to that TIOT, hopefully we get some good discussion!

To clarify too, I don't plan to champion `kable` and `kableExtra`, which have their own annoying quirks. I would love to copy and paste a nicely formatted table straight out of word report into manuscript! I just want to make sure that `flextable` has some of the features we typically use. 

 - per slack today: I'd like to be sure that there is a workaround for long tables in flextable that don't repeat the latex label and cause the page number bug
 - https://github.com/FredHutch/Appel896Analysis/blob/main/Appel896_PK_Report/Appel896_PK_Report.pdf
    - Tables 4 and 6: just making sure all of the greek letters and math symbols render correctly in any location on the table, and use of the `grouped_rows` functionality

That report came to mind quickly for me, but I'll need to think about other examples. I can post them here when I stumble across them

**Comment by kelliemac** [comment]
thanks @mayerbry! I will check on those elements sometime next week to see how flextable handles them. and yeah just post more thoughts here about important functionality as you think of it and I'll add that to the queue!

sidenote: I think it will be good anyways to have these examples in the skeleton.Rmd file's example table so that that example is more comprehensive.

**Comment by slager** [comment]
`flextable` depends on `textshaping`, which is tricky to install on `statsrv`. To install `flextable`, first install `textshaping` following these instructions, and then install `flextable`.  This method worked great for me:

=-=-=

Hi Dave,

I hope all is well. I was able to test installing the R package 'textshaping'.

To get the package to install on your local R environment you will have to do a few things to get it working. 


1.) export the PKG_CPPFLAGS_SITE (export PKG_CPPFLAGS_SITE="-I/usr/include/fribidi -I/usr/include/harfbuzz")

2.) Once you are within your R shell you will want to run: options(repos='https://p3m.dev/cran/2021-02-15'); You could try a more recent date but we can't assure that the newer version will work in our environment. (Dave note: I didn't do this because the most recent version works fine)

3.) run the install.package('textshaping')command 

You will want to make sure to be running all of these commands from admin-os.pc.scharp.org. Doing this will also make sure that the package is installed on statsrv/cronsrv/rstudio-server. 


If you run into any issues or have any additional questions just reach out to me.


Thanks, 

Deon

**Comment by kelliemac** [comment]
Per discussion on 1/7, make sure generally consistent with PDF and then run current version by Solmaz and/or Ollivier and continue!

**Comment by kelliemac** [comment]
And make sure we have good documentation on how the word document template file has been updated and how to update that in the future (can do this within the template file itself)

**Comment by slager** [comment]
Temporarily unmarking myself as "review requested" to clean up my queue.  Please re-request me when it's ready!

**Comment by kelliemac** [comment]
See snapshots from knitting example reports (PDF and Word documents, etc.) during latest CI checks here: https://github.com/FredHutch/VISCtemplates/actions/runs/14047211603?pr=212

**Comment by kelliemac** [comment]
@mayerbry and @slager will you please review this PR when you have a chance?

File: .github/workflows/R-CMD-check.yaml [hunk 1]
```diff
@@ -51,7 +51,8 @@ jobs:
           fancyhdr colortbl soul setspace relsize makecell threeparttable
           threeparttablex environ trimspaces
 
-      - uses: r-lib/actions/setup-r@v2
```
Comment (slager) [inline .github/workflows/R-CMD-check.yaml:54]: Small preference for not hardcoding these workflow section names. The default names already contain the info and are used across many GitHub R repos, so once you're familiar with looking at these standard workflow runs it's almost more confusing to have custom names.

File: .github/workflows/R-CMD-check.yaml [hunk 2]
```diff
@@ -71,10 +72,11 @@ jobs:
           R -q -e 'remotes::install_version("Hmisc", "4.5-0")'
           R -q -e 'remotes::install_github("FredHutch/VISCfunctions", dependencies = TRUE)'
           R -q -e 'remotes::install_github("FredHutch/VISCtemplates", dependencies = TRUE)'
-          R -q -e 'utils::install.packages(c("ggplot2", "dplyr"))'
+          R -q -e 'utils::install.packages(c("ggplot2", "dplyr", "flextable"))'
           R -q -e 'utils::install.packages("rcmdcheck")'
 
-      - uses: r-lib/actions/setup-r-dependencies@v2
+      - name: Install other R dependencies
```
Comment (slager) [inline .github/workflows/R-CMD-check.yaml:78]: See above comment

File: inst/templates/visc_report/skeleton/skeleton.Rmd [hunk 3]
```diff
@@ -23,8 +25,8 @@ lot: true
 toc: true
 dropVISClogo: false
 output: 
-  VISCtemplates::visc_pdf_document: default
   VISCtemplates::visc_word_document: default
+  VISCtemplates::visc_pdf_document: default
```
Comment (slager) [inline inst/templates/visc_report/skeleton/skeleton.Rmd:29]: Does flipping this order change any behaviors?

**Review by slager — APPROVED** [review]
My review on this was pretty light since I don't use the docx outputs and haven't done much development work on it.  I looked at the Bama docx report and had the following comments, which might also apply to some of the other docx reports:

Figure 1 - suggest horizontal centering
Table 2 - suggest landscape or smaller font size.  Columns are very narrow with text wrap happening over 6+ lines.
Table 4-6 have inconsistent formatting among them, with font size and gridlines varying; some may need landscape or smaller font size.

Overall the word documents look like they'd be usable with a bit of manual tweaking, though, so I think this is an improvement.

File: inst/templates/visc_report/skeleton/skeleton.Rmd (hunk 3 above)
Comment (kelliemac) [reply inst/templates/visc_report/skeleton/skeleton.Rmd:29]: this just happens by default when you click the "knit" button in rstudio. I will revert this.

File: .github/workflows/R-CMD-check.yaml (hunk 1 above)
Comment (kelliemac) [reply .github/workflows/R-CMD-check.yaml:54]: ok, I will revert! I don't feel strongly about this.

**Comment by kelliemac** [comment]
> I looked at the Bama docx report and had the following comments, which might also apply to some of the other docx reports:
> 
> Figure 1 - suggest horizontal centering Table 2 - suggest landscape or smaller font size. Columns are very narrow with text wrap happening over 6+ lines. Table 4-6 have inconsistent formatting among them, with font size and gridlines varying; some may need landscape or smaller font size.
> 
> Overall the word documents look like they'd be usable with a bit of manual tweaking, so this definitely seems like an improvement!

@slager I actually don't think there is a centering issue with Figure 1, the ggplot figures just look off-set because of the space in the left margin for the y-axis label. unless I am missing something?

and regarding the table formatting, this is a limitation of using `kable()` and `kableExtra`. you can see that the versions of tables produced with `flextable` are formatted nicely. I didn't want to remove the `kable()` calls entirely from the skeleton because I don't want to force people to use `flextable`, but I think it's nice to show how much better the formatting is in the Word output and also what the differences are in the pdf output... @mayerbry how do you think we should handle this? one option is to have conditionals based on the output type (i.e., use kable for pdf and flextable for word), but that can get a bit unwieldy across a full report. I think it would be fine for the skeleton but I am hesitant to encourage people to do that all the time.

**Comment by slager** [comment]
Oh, I see!  I overlooked the fact that you were purposely using kable and flextable one after the other there.  That all makes sense then.

**Comment by kelliemac** [comment]
@slager good feedback, I don't think it was that obvious. I went ahead and updated the two `skeleton.Rmd` files to include better comments/text explaining why there are duplicate tables.

**Comment by kelliemac** [comment]
update: per conversation with brian and for a better solution to dave's question about why the kable tables were poorly formatted in the Word doc output, I edited the skeleton.Rmd files to only output kable OR flextable tables, depending on whether `output_type == 'latex'`. now the skeleton files knitted to either Word or PDF look nice!!!

one outstanding issue (that I don't think we should address in this PR but I can make a separate issue for later) is: format of study schema table in Word doc output is not so nice. changing this will require changing the `study_schema()` function, though, and that is a can of worms I think we should address separately.

@mayerbry when you have a chance can you please review and let me know if you're happy with everything? once you've done that I can merge this PR!

**Comment by kelliemac** [comment]
@mayerbry can you take one more look at this PR this week? I think it is ready to be merged.

**Review by mayerbry — APPROVED** [review]
